File: examples/kotlin/examples/ClientSimpleCrud.kt
package io.ekodb.client.examples

import io.ekodb.client.*
import io.ekodb.client.types.FieldType
import io.ekodb.client.types.Record
import kotlinx.coroutines.runBlocking
import java.util.*

/**
 * Simple CRUD operations example
 */
fun main() = runBlocking {
    val baseUrl = System.getenv("API_BASE_URL") ?: "http://localhost:8080"
    val apiKey = System.getenv("API_BASE_KEY") ?: error("API_BASE_KEY environment variable not set")
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_users_example"
    
    println("=== ekoDB Kotlin Client - Simple CRUD Example ===\n")
    
    try {
        // Create - Insert document with various field types
        println("=== Create ===")
        val user = Record.new()
            .insert("name", "Alice Johnson")
            .insert("email", "alice@example.com")
            .insert("age", 28)
            .insert("active", true)
            .insert("price", 99.99)
            .insert("created_at", Date().toString())
            .insert("user_id", "550e8400-e29b-41d4-a716-446655440000")
            .insert("tags", FieldType.array(listOf(
                FieldType.string("tag1"),
                FieldType.string("tag2"),
                FieldType.string("tag3")
            )))
            .insert("metadata", FieldType.obj(mapOf(
                "key" to FieldType.string("value"),
                "nested" to FieldType.obj(mapOf("deep" to FieldType.boolean(true)))
            )))
            .insert("embedding", FieldType.vector(listOf(0.1, 0.2, 0.3, 0.4, 0.5)))
            .insert("categories", FieldType.set(listOf(
                FieldType.string("electronics"),
                FieldType.string("computers")
            )))
            .insert("data", FieldType.binary(java.util.Base64.getDecoder().decode("aGVsbG8gd29ybGQ=")))
        
        val inserted = client.insert(collection, user)
        println("✓ Inserted user: $inserted")
        
        val userId = (inserted["id"] as? FieldType.StringValue)?.value
            ?: error("No ID returned")
        println("  User ID: $userId\n")
        
        // Read
        println("=== Read ===")
        val found = client.findById(collection, userId)
        println("✓ Found user by ID: $found\n")
        
        // Read - Extract field values using type-specific getValue utilities
        println("=== Extract Field Values (All Types) ===")
        // ekoDB returns fields as Map with "type" and "value"
        // Use type-specific getValue functions to extract values
        val name = getValue<String>(found["name"])
        val email = getValue<String>(found["email"])
        val age = getValue<Int>(found["age"])
        val active = getValue<Boolean>(found["active"])
        val price = getDecimalValue(found["price"])
        val createdAt = getDateTimeValue(found["created_at"])
        val userIdField = getUUIDValue(found["user_id"])
        val tags = getArrayValue(found["tags"])
        val metadata = getObjectValue(found["metadata"])
        val embedding = getVectorValue(found["embedding"])
        val categories = getSetValue(found["categories"])
        val data = getBytesValue(found["data"])
        
        println("Extracted values:")
        println("  name (String): $name")
        println("  email (String): $email")
        println("  age (Integer): $age")
        println("  active (Boolean): $active")
        println("  price (Decimal): $price")
        println("  created_at (DateTime): $createdAt")
        println("  user_id (UUID): $userIdField")
        println("  tags (Array): $tags")
        println("  metadata (Object): $metadata")
        println("  embedding (Vector): $embedding")
        println("  categories (Set): $categories")
        println("  data (Bytes): ${data?.size ?: 0} bytes")
        
        // Record already contains all fields
        println("Record fields: ${found.keys.joinToString(", ")}\n")
        
        // Update
        println("=== Update ===")
        val updates = Record.new()
            .insert("age", 29)
            .insert("city", "San Francisco")
        
        val updated = client.update(collection, userId, updates)
        println("✓ Updated user: $updated\n")
        
        // Query
        println("=== Query ===")
        val query = QueryBuilder.new()
            .gte("age", 25)
            .eq("active", true)
            .limit(10)
            .build()
        
        val results = client.find(collection, query)
        println("✓ Found ${results.size} users matching query")
        results.forEach { println("  - $it") }
        println()
        
        // Delete
        println("=== Delete ===")
        client.delete(collection, userId)
        println("✓ Deleted user with ID: $userId\n")
        
        // Verify deletion
        try {
            client.findById(collection, userId)
            println("⚠ User still exists (unexpected)")
        } catch (e: Exception) {
            println("✓ Confirmed user was deleted\n")
        }
        
    } finally {
        // Cleanup
        println("=== Cleanup ===")
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n=== Example Complete ===")
    }
}

File: examples/rust/examples/client_simple_crud.rs
//! Simple CRUD operations using ekodb_client library
//!
//! This example demonstrates the same operations as simple_crud.rs
//! but using the ekodb_client library for cleaner, type-safe code.
//!
//! Compare with simple_crud.rs to see the difference!

use ekodb_client::{Client, FieldType, Record};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Create ekoDB client (automatically handles token exchange)
    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("✓ Client created (token exchange happens automatically)");

    let collection = "client_simple_crud_rust";

    // Example 1: Insert a document with various field types
    println!("\n=== Insert Document ===");
    let mut record = Record::new();
    record.insert("name", "Test Record");
    record.insert("value", 42);
    record.insert("active", true);
    record.insert("price", 99.99);
    record.insert("created_at", chrono::Utc::now().to_rfc3339());
    record.insert("user_id", "550e8400-e29b-41d4-a716-446655440000");
    record.insert(
        "tags",
        FieldType::array(vec![
            FieldType::string("tag1"),
            FieldType::string("tag2"),
            FieldType::string("tag3"),
        ]),
    );
    record.insert(
        "metadata",
        serde_json::json!({"key": "value", "nested": {"deep": true}}),
    );
    record.insert(
        "embedding",
        FieldType::vector(vec![0.1, 0.2, 0.3, 0.4, 0.5]),
    );
    record.insert(
        "categories",
        FieldType::set(vec![
            FieldType::string("electronics"),
            FieldType::string("computers"),
        ]),
    );
    record.insert(
        "data",
        base64::Engine::encode(&base64::engine::general_purpose::STANDARD, b"hello world"),
    );

    let inserted = client.insert(collection, record, None).await?;
    println!("Inserted: {:?}", inserted);

    // Extract the ID
    let doc_id = if let Some(id_field) = inserted.get("id") {
        match id_field {
            FieldType::String(id) => id.clone(),
            _ => return Err("Expected string ID".into()),
        }
    } else {
        return Err("No ID returned".into());
    };

    // Example 2: Find by ID
    println!("\n=== Find by ID ===");
    let found_doc = client.find_by_id(collection, &doc_id, None).await?;
    println!("Found: {:?}", found_doc);

    // Example 2b: Extract field values using type-specific getValue utilities
    println!("\n=== Extract Field Values (All Types) ===");
    // ekoDB returns fields as FieldType enum
    // For direct extraction, we can pattern match on FieldType
    let name_value = match found_doc.get("name") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::String(s)) => Some(s.clone()),
            _ => None,
        },
        Some(FieldType::String(s)) => Some(s.clone()),
        _ => None,
    };
    let value_value = match found_doc.get("value") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::Integer(i)) => Some(*i),
            _ => None,
        },
        Some(FieldType::Integer(i)) => Some(*i),
        _ => None,
    };
    let active_value = match found_doc.get("active") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::Boolean(b)) => Some(*b),
            _ => None,
        },
        Some(FieldType::Boolean(b)) => Some(*b),
        _ => None,
    };
    let price_value = match found_doc.get("price") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::Float(f)) => Some(*f),
            _ => None,
        },
        Some(FieldType::Float(f)) => Some(*f),
        _ => None,
    };
    let created_at_value = match found_doc.get("created_at") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::String(s)) => Some(s.clone()),
            Some(FieldType::DateTime(dt)) => Some(dt.to_rfc3339()),
            _ => None,
        },
        Some(FieldType::String(s)) => Some(s.clone()),
        Some(FieldType::DateTime(dt)) => Some(dt.to_rfc3339()),
        _ => None,
    };
    let user_id_value = match found_doc.get("user_id") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::String(s)) => Some(s.clone()),
            _ => None,
        },
        Some(FieldType::String(s)) => Some(s.clone()),
        _ => None,
    };
    let tags_value = match found_doc.get("tags") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::Array(arr)) => Some(arr.clone()),
            _ => None,
        },
        Some(FieldType::Array(arr)) => Some(arr.clone()),
        _ => None,
    };
    let metadata_value = match found_doc.get("metadata") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::Object(obj)) => Some(obj.clone()),
            _ => Some(map.clone()),
        },
        _ => None,
    };
    let embedding_value = match found_doc.get("embedding") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::Array(arr)) | Some(FieldType::Vector(arr)) => Some(arr.clone()),
            _ => None,
        },
        Some(FieldType::Vector(vec)) => Some(vec.clone()),
        _ => None,
    };
    let categories_value = match found_doc.get("categories") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::Array(arr)) | Some(FieldType::Set(arr)) => Some(arr.clone()),
            _ => None,
        },
        Some(FieldType::Set(set)) => Some(set.clone()),
        _ => None,
    };
    let data_value = match found_doc.get("data") {
        Some(FieldType::Object(map)) => match map.get("value") {
            Some(FieldType::String(s)) => {
                base64::Engine::decode(&base64::engine::general_purpose::STANDARD, s).ok()
            }
            Some(FieldType::Bytes(b)) => Some(b.clone()),
            _ => None,
        },
        Some(FieldType::String(s)) => {
            base64::Engine::decode(&base64::engine::general_purpose::STANDARD, s).ok()
        }
        Some(FieldType::Bytes(b)) => Some(b.clone()),
        _ => None,
    };

    println!("Extracted values:");
    println!("  name (String): {:?}", name_value);
    println!("  value (Integer): {:?}", value_value);
    println!("  active (Boolean): {:?}", active_value);
    println!("  price (Decimal): {:?}", price_value);
    println!("  created_at (DateTime): {:?}", created_at_value);
    println!("  user_id (UUID): {:?}", user_id_value);
    println!(
        "  tags (Array): {} items",
        tags_value.as_ref().map(|v| v.len()).unwrap_or(0)
    );
    println!(
        "  metadata (Object): {} keys",
        metadata_value.as_ref().map(|v| v.len()).unwrap_or(0)
    );
    println!(
        "  embedding (Vector): {} dims",
        embedding_value.as_ref().map(|v| v.len()).unwrap_or(0)
    );
    println!(
        "  categories (Set): {} items",
        categories_value.as_ref().map(|v| v.len()).unwrap_or(0)
    );
    println!(
        "  data (Bytes): {} bytes",
        data_value.as_ref().map(|v| v.len()).unwrap_or(0)
    );

    // Example 3: Find with query
    println!("\n=== Find with Query ===");
    use ekodb_client::Query;

    let query = Query::new()
        .filter(serde_json::json!({
            "type": "Condition",
            "content": {
                "field": "active",
                "operator": "Eq",
                "value": true
            }
        }))
        .limit(10);

    let docs = client.find(collection, query, None).await?;
    println!("Found documents: {:?}", docs);

    // Example 4: Update document
    println!("\n=== Update Document ===");
    let mut updates = Record::new();
    updates.insert("name", "Updated Record");
    updates.insert("value", 100);

    let updated = client.update(collection, &doc_id, updates, None).await?;
    println!("Updated: {:?}", updated);

    // Example 5: Delete document
    println!("\n=== Delete Document ===");
    client.delete(collection, &doc_id, None).await?;
    println!("Deleted document");

    // Cleanup: Delete the collection
    println!("\n=== Cleanup ===");
    client.delete_collection(collection).await?;
    println!("✓ Deleted collection");

    println!("\n✓ All CRUD operations completed successfully");

    Ok(())
}
